### ISSUE TITLE
Address index, tag and offset

### ISSUE THREAD

**Issue by tmsimont** [tmsimont/cache-simulator#1]
The Address class looks like it has the [index, tag and offset properties on it](https://github.com/tmsimont/cache-simulator/blob/master/trace/Address.cc#L20)

This was a mistake I made in the UML.

The Cache needs to tell you what the index, tag and offset of an address are, because L1 and L2 will have different values for these properties for the same address. So, the address itself can't know what it's tag, offset or index is because these properties are specific to each cache.

For example, if L1 is 2-way associative, it only needs 1 bit for index.

If L2 is 4-way associative in the same architecture, then it will have to use 2 bits of the address to index the set.

This would bean that the tag size is different for the two caches is the same architecture.

So I think we need 

`std::bitset Cache::getTag(Address a);`

 for each of these three properties

Sorry to generate confusion with the bad UML

**Comment by tmsimont** [comment]
FYI --

In the JavaScript layer, I'm doing the following on Cache construction:

```
  this.indexSize = Math.log2(parameters.associativity);
  this.offsetSize = Math.log2(parameters.blockSize);
  this.tagSize = parameters.addressSize - this.indexSize - this.offsetSize;

```

So each Cache instance knows it needs (log2 of associativity) bits for index addressing, (log2 of blockSize) for offset addressing, and the rest of the address goes to tag.

These values can then be used to parse out the bits of the address that point to index, offset and tag:

```
Cache.prototype.getIndexBits = function(addressBits) {
  return addressBits.slice(
    this.tagSize, this.tagSize + this.indexSize
  );
}
Cache.prototype.getIndex = function(addressBits) {
  return parseInt(this.getIndexBits(addressBits), 2)
}

Cache.prototype.getTagBits = function(addressBits) {
  return addressBits.slice(0, this.tagSize);
}
Cache.prototype.getTag = function(addressBits) {
  return parseInt(this.getTagBits(addressBits), 2)
}

Cache.prototype.getOffsetBits = function(addressBits) {
  return addressBits.slice(
    this.tagSize + this.indexSize,
    this.addressSize
  );
}
Cache.prototype.getOffset = function(addressBits) {
  return parseInt(this.getOffsetBits(addressBits), 2)
}

```

`slice()` here just returns the part of the address bits from the first argument position to the second argument position.

The address split is:

```
_______________________________________
|1010101001010101010101... for 32 bits |
|        tag     |  index |   offset   |     
---------------------------------------
```

**Comment by tmsimont** [comment]
Oops I made an error....

The number of index bits should be:

```
 this.indexSize = Math.log2(NUMBER_OF_BLOCKS / parameters.associativity);
```

or just log2(number of sets) -- I was incorrectly calculating number of sets based on associativity

**Comment by tmsimont** [comment]
Fixed in commit [1b7482e](https://github.com/tmsimont/cache-simulator/commit/1b7482e4da368fbf663cc3cc6969dc0e38dc6272)
